src/Microsoft.AspNetCore.Hosting/WebHostBuilder.cs

```diff
@@ -201,6 +201,8 @@ private IServiceCollection BuildHostingServices()
             // Conjure up a RequestServices
             services.AddTransient<IStartupFilter, AutoRequestServicesStartupFilter>();
 
+            services.AddTransient<IServiceProviderFactory, DefaultServiceProviderFactory>();
```

Comment:
Yes this should do a TryAdd instead of an add. 

> Introduce a UseServiceProviderFactory method on the IWebHostBuilder interface, like I've done here

I considered this, the only question is do we want to let `IServiceProviderFactory` implementations rely on services? It's probably a bad idea anyways but you can imagine this relying on `IOptions<T>` configured in the `WebHostBuilder`.

I think I'll make it a field instead.